File: integration/matrix_con.py
import sys
import os

# Add the toolbox to path so we can import it
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "python"))

from microservice_toolbox.serializers.providers import new_bin_serializer, new_json_serializer

def main():
    format = "msgpack"
    if len(sys.argv) > 1:
        format = sys.argv[1]

    if format == "json":
        s = new_json_serializer()
    else:
        s = new_bin_serializer()
    
    # Read binary data from stdin
    data = sys.stdin.buffer.read()
    
    try:
        decoded = s.unmarshal(data, dict)
        expected = {"name": "Integration", "value": 100}
        
        if decoded == expected:
            sys.stderr.write(f"Python: Success ({format})\n")
            sys.exit(0)
        else:
            sys.stderr.write(f"Python: Data mismatch ({format}). Got {decoded}\n")
            sys.exit(1)
    except Exception as e:
        sys.stderr.write(f"Python: Error during deserialization ({format}): {e}\n")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: integration/matrix_con.go
package main

import (
	"fmt"
	"io"
	"os"

	"github.com/Bastien-Antigravity/microservice-toolbox/go/pkg/serializers"
)

type IntegrationData struct {
	Name  string `json:"name" msgpack:"name"`
	Value int    `json:"value" msgpack:"value"`
}

func main() {
	format := "msgpack"
	if len(os.Args) > 1 {
		format = os.Args[1]
	}

	var s serializers.Serializer
	if format == "json" {
		s = serializers.NewJSONSerializer()
	} else {
		s = serializers.NewBinSerializer()
	}

	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to read stdin: %v\n", err)
		os.Exit(1)
	}

	var decoded IntegrationData
	err = s.Unmarshal(data, &decoded)
	if err != nil {
		fmt.Fprintf(os.Stderr, "unmarshal error: %v\n", err)
		os.Exit(1)
	}

	expected := IntegrationData{Name: "Integration", Value: 100}
	if decoded.Name == expected.Name && decoded.Value == expected.Value {
		fmt.Fprintf(os.Stderr, "Go: Success (%s)\n", format)
		os.Exit(0)
	} else {
		fmt.Fprintf(os.Stderr, "Go: Data mismatch (%s). Got %+v\n", format, decoded)
		os.Exit(1)
	}
}
